### PR TITLE
Timely report checks

### PR DESCRIPTION
**Pull request by zewolfe**
Checks whether a timely report has been sent before and whether the current report being sent is timely.

### CODE REVIEW

File: dhis2-android-app/src/main/java/org/dhis2/mobile/utils/IsTimely.java [hunk 1]
```diff
@@ -14,16 +19,32 @@
     public IsTimely(){
 
     }
-    public static Boolean check(int period){
+    public static Boolean check(String period){
```
Comment (quesebifurcan) [inline dhis2-android-app/src/main/java/org/dhis2/mobile/utils/IsTimely.java]: Can we add a couple of tests for this method?

The problem is that it's written in such a way that it returns different results at different times (i.e. it is not referentially transparent), so you either have to

1. split it out into two methods, one of which takes two arguments: (calendar, periodString). This method can then be tested easily.
2. Mock Calendar to simulate different points in time.

In summary:

- [ ] Refactor / Mock
- [ ] Add tests for timely period strings
- [ ] Add tests for untimely period strings

File: dhis2-android-app/src/main/java/org/dhis2/mobile/utils/IsTimely.java [hunk 2]
```diff
@@ -14,16 +19,32 @@
...
+        int year = Calendar.getInstance().get(Calendar.YEAR);
+
 
-        if(period == weekNumber -2 && currentDay == Calendar.MONDAY && currentHour < 12){
+        if(!(periodYear < year) &&  periodWeek == weekNumber -2 && currentDay == Calendar.MONDAY && currentHour < 12){
             timely = true;
         }
         return timely;
     }
+    public static Boolean hasBeenSet(ArrayList<Group> groups){
```
Comment (quesebifurcan) [inline dhis2-android-app/src/main/java/org/dhis2/mobile/utils/IsTimely.java]: Can we add a test for this method as well?

**Comment by quesebifurcan** [comment]
+1

File: dhis2-android-app/src/test/java/IsTimelyTest.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+import static org.junit.Assert.assertThat;
+import static org.mockito.Mockito.when;
+
+/**
+ * Created by george on 10/4/16.
+ */
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({IsTimely.class})
+public class IsTimelyTest {
+    private final String period = "2016W37";
```
Comment (quesebifurcan) [inline dhis2-android-app/src/test/java/IsTimelyTest.java]: What happens when this period string is (for example): `2017W01`?

I tried modifying this file to create a passing test for such a date, but there are some comparisons of integers here: https://github.com/eHealthAfrica/dhis2-android-datacapture/pull/17/files#diff-fdacc0de5c4cfc79cddf0002501b3253R31 which complicate things (what is `weekNumber - 2` if `weekNumber` is `1`?)

File: dhis2-android-app/src/main/java/org/dhis2/mobile/utils/IsTimely.java [hunk 4]
```diff
@@ -14,16 +19,32 @@
...
+        int currentDay = calendar.get(Calendar.DAY_OF_WEEK);
+        int currentHour = calendar.get(Calendar.HOUR_OF_DAY);
+        int year = calendar.get(Calendar.YEAR);
 
-        int weekNumber = Calendar.getInstance().get(Calendar.WEEK_OF_YEAR);
-        int currentDay = Calendar.getInstance().get(Calendar.DAY_OF_WEEK);
-        int currentHour = Calendar.getInstance().get(Calendar.HOUR_OF_DAY);
-
-        if(period == weekNumber -2 && currentDay == Calendar.MONDAY && currentHour < 12){
+        if(!(periodYear < year) &&  periodWeek == weekNumber -2 && currentDay == Calendar.MONDAY && currentHour < 12){
```
Comment (quesebifurcan) [inline dhis2-android-app/src/main/java/org/dhis2/mobile/utils/IsTimely.java]: See comment below, we should probably use proper time deltas instead of integers here.

File: dhis2-android-app/src/main/java/org/dhis2/mobile/utils/IsTimely.java [hunk 5]
```diff
@@ -14,16 +19,34 @@
...
+        int currentHour = calendar.get(Calendar.HOUR_OF_DAY);
+        int year = calendar.get(Calendar.YEAR);
 
-        int weekNumber = Calendar.getInstance().get(Calendar.WEEK_OF_YEAR);
-        int currentDay = Calendar.getInstance().get(Calendar.DAY_OF_WEEK);
-        int currentHour = Calendar.getInstance().get(Calendar.HOUR_OF_DAY);
-
-        if(period == weekNumber -2 && currentDay == Calendar.MONDAY && currentHour < 12){
+        if(!(periodYear < year) &&
+                periodWeek == (weekNumber -2 < 0 ? 52 : weekNumber-2) &&
```
Comment (quesebifurcan) [inline dhis2-android-app/src/main/java/org/dhis2/mobile/utils/IsTimely.java]: This doesn't work, it seems. 

- [ ] The `timelyCheckShouldReturnTrue()` passes for period string `2016W37`, but it also passes for `2345W37`, which is perhaps not what we want.

- [ ] The method returns false if `periodYear` is lower than `year.` But if a report from the last week of 2016 is to be approved at the beginning of 2017, the years will be different and the report might still be "timely"

- [ ] If we are going to use the period strings (are there other options?), then we should at least use multiple period strings in the test.

- [ ] Are the weeks you're using with `Calendar` actually the same ISO weeks that DHIS2 is using internally?

File: dhis2-android-app/src/test/java/IsTimelyTest.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
```
Comment (quesebifurcan) [inline dhis2-android-app/src/test/java/IsTimelyTest.java]: It would be nice to have tests for other strings than `2016W37` (see above).

**Comment by quesebifurcan** [comment]
+1
